File: computer_science/data_structures/stack/stack.cpp
#include <iostream>

using namespace std;

#define MAX_STACK_SIZE 10

class Stack {
private:
	int container[MAX_STACK_SIZE];
	int numberOfElements;

public:
	Stack() {
		numberOfElements = 0;
	}

	void push(int element) {
		if (numberOfElements > MAX_STACK_SIZE) {
			cout << "Error: Stack Overflow";
		} else {
			container[numberOfElements] = element;
			numberOfElements++;
		}
	}

	int pop() {
		if (isEmpty()) {
			cout << "Error: No element to pop";
		} else {
			numberOfElements--;
			return container[numberOfElements];
		}
	}

	int top() {
		return container[numberOfElements-1];
	}

	bool isEmpty() {
		return numberOfElements == 0;
	}
};

int main() {
	Stack stack;

	if (stack.isEmpty()) cout << "Stack is empty" << endl;

	stack.push(1);
	stack.push(2);
	stack.push(3);

	cout << stack.pop() << endl;

	stack.push(10);

	cout << stack.top() << endl;

	if (!stack.isEmpty()) cout << "Stack is not empty" << endl;

	return 0;
}

File: coding_interviews/javascript/stack/stack.js
let stack = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

function push(stack, value) {
  stack.push(value);
}

function pop(stack) {
  return stack.pop();
}

function top(stack) {
  return stack[stack.length - 1];
}

function isEmpty(stack) {
  return stack.length === 0;
}

console.log(stack); // [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

push(stack, 11);
console.log(stack); // [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

pop(stack); // 11
console.log(stack); // [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

top(stack); // 10
isEmpty(stack); // false

while (!isEmpty(stack)) {
  stack.pop();
}

isEmpty(stack); // true
console.log(stack);
